### ISSUE TITLE
Invalid prop: type check failed for prop "data". Expected Object, got Array.

### ISSUE THREAD

**Issue by dainemedia** [gilbitron/laravel-vue-pagination#18]
When passing a Laravel paginator response to the data prop of this component I receive the following error:

    Invalid prop: type check failed for prop "data". Expected Object, got Array.

The problem is, the data passed to the data prop is, in fact, an object and is indicated as so in Vue DevTools.

![Screenshot](https://image.ibb.co/cXCASx/Screen_Shot_2018_03_14_at_11_26_41.png)

As you can see from the screenshot above, it is indeed an object, as expected by the component and indicated in the docs.

I'm using Vue 2.

Unless the type check is on the data property of the paginator response, then something is going wrong here.

That said, in the docs, it gives an example of the type of response required, and the data property of the paginator response is an array:

    {
	    current_page: 1,
	    data: [],
	    from: 1,
	    last_page: 1,
	    next_page_url: null,
	    per_page: 10,
	    prev_page_url: null,
	    to: 1,
	    total: 0,
    }

**Comment by GitzJoey** [comment]
understand that this only a [Vue-Warn] but its would be nice if it can compatible with the laravel json response (without any warn shown)

**Comment by gilbitron** [comment]
I've just tested this with the latest version of Laravel and couldn't recreate the issue. I've [updated the README.md](https://github.com/gilbitron/laravel-vue-pagination/blob/f782a14e987f4eadefd6d1595251cbd32844d6c5/README.md) to be more in line with current Vue/Laravel standards (i.e. using axios instead of vue-resource). Maybe having a look at the refreshed README.md might help?

The only things I can think of that might be causing this issue: 

* If you are using axios, you need to return `response.data` and not just the `response` itself.
* Make sure you are not returning the `data` array from the paginator response instead of the whole response object.

```js
getResults(page = 1) {
    axios.get('example/results?page=' + page)
        .then(response => {
            this.laravelData = response.data;
        });
    }
```

**Comment by GitzJoey** [comment]
I'm using the latest laravel and also using the axios, you can check in my project
https://github.com/GitzJoey/TKBARU/blob/master/resources/assets/js/apps/supplier.js

I can provide you a test pen in my project github if want

**Comment by abr4xas** [comment]
I have this problem too...

This is my response: 

```javascript
{
    current_page: 1, 
    data: Array(10), 
    first_page_url: "http://localhost:8000/admin/post?page=1", 
    from: 1, 
    last_page: 17,
    ...
}
```

Error:

```javascript
[Vue warn]: Invalid prop: type check failed for prop "data". Expected Object, got Array.
```

Btw: 

With laravel 5.6.x

```javascript
getResults(page = 1) {
    axios.get('example/?page=' + page) // <-- this 
        .then(response => {
            this.laravelData = response.data;
        });
}
```

**Comment by sunscreem** [comment]
I had the same problem. 

I solved it by making sure I changed this:

`<li v-for="post in laravelData" ...`

to this:

`<li v-for="post in laravelData.data" ...`

**Comment by GitzJoey** [comment]
nope. still showing the warn
i tried

laravelData : Object
laravelData.data : Object

laravelData : Array
laravelData.data : Object

**Comment by sunscreem** [comment]
Can you share your code, maybe we can spot that issue for you?

**Comment by GitzJoey** [comment]
@sunscreem 
https://github.com/GitzJoey/TKBARU/blob/master/resources/assets/js/apps/supplier.js

note: its standard live laravel project also if you care to run it (i can guide you to run it also)

**Comment by sunscreem** [comment]
@GitzJoey You see here:

https://github.com/GitzJoey/TKBARU/blob/c26742fae6fc15533d01c00ecabe3e471aa7f104/resources/assets/js/apps/supplier.js#L62

Can you show me a console.log on `this.supplierList` please?

**Comment by gilbitron** [comment]
@ettiennelouw Good shout. I've raised a separate issue for supporting Laravel's API Resources #22.

**Comment by GitzJoey** [comment]
@sunscreem 
hi, sorry for late response, here;s the console.log on line 62
<img width="740" alt="capture" src="https://user-images.githubusercontent.com/5994569/41758027-ff1683ec-760f-11e8-80da-1985a111dbd1.PNG">

**Comment by Hujjat** [comment]
It's because your ` data ` is an array. I should be an object. 

Yours is like this 

data : [10]

It should be 

data : {}

For me, it fixed the warning.

**Comment by dieterc** [comment]
> I had the same problem.
> 
> I solved it by making sure I changed this:
> 
> `<li v-for="post in laravelData" ...`
> 
> to this:
> 
> `<li v-for="post in laravelData.data" ...`

This was also the fix for me

**Comment by nazaninhesari** [comment]
not work in laravel 5.5 
larvael return default type of paginate , how i can convert it to object ?

**Comment by BoraVivek** [comment]
Make sure you are setting the data in variable as object instead of array 

If you have data.data in http call:

`axios.get("api/post").then(({ data }) => (this.posts = data.data));`

Change it to :

`axios.get("api/post").then(({ data }) => (this.posts = data));`

and in loop

if you have this :

`<tr v-for="post in posts" :key="post.id">`

add the data here like this :

`<tr v-for="post in posts.data" :key="post.id">`

**Comment by ManojKiranA** [comment]
> try to change initial value in data() from [] to {}.

its worked like a charm

**Comment by AliN11** [comment]
> try to change initial value in data() from [] to {}.

How do you do that properly?

**Comment by cvhainb** [comment]
I have the same error. I have solved with this.

```
<div v-if="posts.total > 0">
    <pagination :data="posts" @pagination-change-page="getPaginationResults"></pagination>
</div>
```

**Comment by rahmanrezaee** [comment]
1: instead []  use {}
2:
`getResults(page = 1) {
    axios.get('example/?page=' + page) // <-- this 
        .then(response => {
            this.laravelData = response.data;
        });
}
`
3:instead <tr v-for="post in posts" :key="post.id"> us <tr v-for="post in posts.data" :key="post.id">

**Comment by uxweb** [comment]
Just a quick note about this issue:

TLDR; **I'm sure this issue is related on how `json_encode()` works and can be tweaked by passing it some *options* or taking care of the structure of data we send in the response**

I've tested the following scenarios for passing data from a Laravel response to Vue component props:

**Empty PHP Array**:

An empty PHP Array will be converted to a JS empty Array.

```
// PHP
[]

// JS
[]
```

**Associative PHP Array (with string index)**:

An associative PHP Array will be converted to a JS Object.

```
// PHP
[
    'index1' => 'Hello',
    'index2' => [ 'item1', 'item2' ]
]

// JS
{
    index1: 'Hello',
    index2: [ 'item1', 'item2' ]
}
```
> This is the weird case, the prop is set to expect an Array, because that is what you are returning in the Laravel response, but get an object if the array has associative keys, and guess what? If the array is empty you get an array!

**Multidimensional PHP Array (with numeric index)**:

A multidimensional PHP Array with numeric index will be converted to a JS Array

```
// PHP
[
    [ 'name' => 'foo', 'messages' => [ 'message1', 'message2' ] ],
    [ 'name' => 'bar', 'messages' => [ 'message1', 'message2' ] ],
]

// JS
[
    { name: 'foo', messages: [ 'message1', 'message2' ]},
    { name: 'bar', messages: [ 'message1', 'message2' ]},
]
```

**Comment by alphadirectory** [comment]
you are declaring one type of list, and when you consume it from the API it goes over another type.
To fix the problem, you can declare multiple forms of the list as:
props: {
data: [array, object]
}
inthis way, Vuejs will read any list you can give!

**Comment by samoudwfr** [comment]
I had had the same warning, and now it's working perfectly.
juste replace
data: [ ]  by  data: { }

**Comment by mhlanga-joseph** [comment]
@BoraVivek
Answer worked for me. Thanks

**Comment by atefrihane** [comment]
I'm passing a prop from the laravel blade to vue component how can I convert it into an object ?
I'm not using axios..
`  props: ['data'],     // always returning  array`

**Comment by mukramin97** [comment]
[Vue warn]: Error in render: "TypeError: Cannot read property 'data' of undefined"

i have some warning too, i am passing 2 tables from laravel to axios, how to make these 2 data into an object?
i got no problem with other pagination with only 1 table passed

Controller :
public function index()
    {  
        $groupOne = Santri::with('Asrama')->paginate(5);
        $groupTwo = Asrama::all();

        return (compact('groupOne', 'groupTwo'));
    }

Santri.vue:
getResults(page = 1) {
                axios.get('api/santri?page=' + page)
                    .then(response => {
                        this.santris = response.data;
                    });
            },
loadSantri(){
                axios.get("api/santri").then(({ data }) => (this.santris = data ));
                
            },

data(){
            return {
                editMode: false,
                santris : {
                }
            }

**Comment by ebaguma** [comment]
> I had the same issue but i was using a Resource to return my json then it gives you the output in this format instead of the norm.
> 
> ![screenshot 2018-06-15 12 30 49](https://user-images.githubusercontent.com/1754421/41463829-2f3fc970-7098-11e8-9266-76e66ef0e515.jpg)
> 
> **From the Laravel Documentation**
> 
> Data Wrapping And Pagination
> When returning paginated collections in a resource response, Laravel will wrap your resource data in a data key even if the withoutWrapping method has been called. This is because paginated responses always contain meta and links keys with information about the paginator's state:
> 
> https://laravel.com/docs/5.6/eloquent-resources#data-wrapping

I followed this solution to solve my problem by opting to use database views for all cases where i was using resources. Hope that helps someone having similar issues

**Comment by bugohoss12** [comment]
I have solved the issue by wrapping the api response into: Object.assign({}, response);

**Comment by romeldev** [comment]
use the following structure in your vue data:
`dataItems: { data: [] }`

in your paging component set:
`:data = "dataItems"`

and to list your data in a table or html list use:
`v-for="(item, key) in dataItems.data"`

**Comment by Afolorunso** [comment]
### Fixed!

So I was able to use Javascript `Object.fromEntries() ` method to parse the data to an Object since the package is reading it as an array.

**Laravel Controller**
        `return response()->json(["applicants" => $applications], 200);`

**Vue file**
` data: () => ({
applicants:{}
})

//Method
getJobApplicants(){
   this.applicants = response.data.applicants.data;
}

//<div body></div>
 <pagination
          :data="Object.fromEntries(applicants)"
          @pagination-change-page="getJobApplicants"
      >
          <span slot="prev-nav">Previous </span>
          <span slot="next-nav">Next</span>
  </pagination>`
